### ISSUE TITLE
[XProf] Improve support for AMD GPUs

### ISSUE THREAD

**Issue by clarkechong** [ROCm/xla#1106]
# Description

Provide a more native XProf experience for traces on AMD GPUs.
This generally involves changes within XProf and XLA profiler backend.

In short, XLA profiler produces stats in the XSpace protobuf `.xplane.pb` (the native profiling output format of jax.profiler). XProf consumes this XPlane, producing its own internal IR -> data tables, which are then parsed by the frontend.

# Info

- This is subsequent work following https://github.com/ROCm/xla-internal/issues/49
  - Initial survey of current behaviour: https://amd.atlassian.net/wiki/x/IoBqaw
- Related work in progress at https://github.com/ROCm/xla-internal/issues/66
  - theoretical kernel occupancy (which will be used to directly occupy XPlane stat `theoretical_occupancy_pct`)
  - ROCTX annotation enablement, whose effects will be observable within the XProf trace viewer.

# Status

### Upstream
- https://github.com/openxla/xprof/pull/3086 (MERGED)
- https://github.com/openxla/xla/pull/47067 (MERGED)
- https://github.com/openxla/xprof/pull/3163 (MERGED)
- https://github.com/openxla/xprof/pull/3230
- https://github.com/openxla/xla/pull/47917 (MERGED)

### Ongoing
- https://github.com/clarkechong/xprof/pull/2
- https://github.com/clarkechong/xprof/pull/3

### Future work
1. ROCm side for PMC hardware counter sampling system (XLA). NVIDIA already has cupti sampling system implementation (`xla/xla/backends/profiler/gpu/cupti_pm_sampler.h`)

2. Step markers in XPlane (current no step markers, so Overview page is broken)  - ideally a step is inferred by XLA, usually referring to a single module. So a step marker is placed before and after each module (including associated overhead, to measure attribution to idle time/overhead)
  a) so XLA needs to produce step markers either into kernel details or as a new XLine in the device XPlane, for XProf to consume.
  b) this is separate to ROCTX annotations which are user-driven (require a manually inserted AnnotationTrace() in the source)

3. More accurate roofline calculations based off data types within op/kernel (see [this comment](https://github.com/openxla/xprof/pull/3163#issuecomment-5293764385) for context)

**Comment by clarkechong** [comment]
# Pending Work (9/9/2026)

### 1. Improve standalone memory-event kind

**Issue:** ROCm standalone copy/memset details use `kind:Unknown`, although coarse transfer direction and source/destination agents are available. Command-buffer copies usually appear as kernels, so this fix will not cover the dominant graph-captured path.

<img width="1570" height="1225" alt="Image" src="https://github.com/user-attachments/assets/1c58b5c4-2988-4ecf-92b2-aa15ea769b10" />

**Proposal:** Emit the reliable coarse memory/agent kind for standalone events. Use the graph-node attribution described below—not string heuristics—for command-buffer copies.

**Desired behaviour:** Standalone transfers report every memory kind that ROCprofiler can determine reliably and leave finer-grained unknowns explicit.

### 2. Expose GPU duty cycle

**Issue:** XProf constructs a GPU `DutyCycleTracker`, but publishes `busy_time_ps`/`idle_time_ps` and consumes the result only on the TPU path.

**Proposal:** Publish the unioned GPU busy/idle times, compute the generic device duty cycle in Overview, and change the frontend label from “TPU Duty Cycle” to a hardware-neutral label.

Do not populate `normalized_time_ps` as part of this work: it means DVFS/default p-state normalization. `core_type` is a TPU enum and should remain unknown.

**Desired behaviour:** GPU Overview reports a device duty cycle consistent with the union of activity across its streams.

### 3. Make the GPU Roofline dtype-aware

This will require fairly involved changes across XProf and XLA.

**Issue:** The GPU compute ceiling uses the architecture's maximum throughput, normally FP16/BF16. Comparing FP32, FP64, TF32, or FP8 work against that single ceiling can materially misstate efficiency.

See this [issue comment](https://github.com/openxla/xprof/pull/3163#issuecomment-5293764385) for context.

**Proposal:**

- Infer operation and accumulation mode from HLO/compiler metadata.
- Use the vendor architecture tables for the corresponding throughput.
- Keep raw model FLOPs separate from any normalized reference-FLOP metric.
- Use collector metadata only where fused/library kernels cannot be inferred.

**Desired behaviour:** Tests cover FP64, FP32, TF32, FP16, BF16, and FP8 on AMD, with unchanged NVIDIA behavior.

### 4. Restore step-based GPU analysis

The following three tasks form one dependency chain.

#### 4.1 Propagate explicit JAX steps to GPU work

**Issue:** `jax.profiler.StepTraceAnnotation` reaches the host XPlane but does not reach ROCm kernels, leaving Overview and Input Pipeline Analysis empty.

<img width="1854" height="847" alt="Image" src="https://github.com/user-attachments/assets/c3bdef8f-631e-45b3-86ff-e73421ecfb55" />

**Capture pattern:**

```python
jax.profiler.start_trace(TRACE_DIR)
for step in range(10, 21):
    with jax.profiler.StepTraceAnnotation("train", step_num=step):
        state = step_fn(state, batch)
jax.block_until_ready(state)
jax.profiler.stop_trace()
```

On an 8× MI300X run, the resulting trace contains:

- 11 `/host:CPU` `train` roots with `_r=1` and the correct `step_num`;
- matching PJRT producer/consumer events;
- 6,072 correlation IDs shared by host launches and device kernels;
- 40,216 kernels across eight GPU planes, but no propagated `group_id`;
- zero step rows after XProf 2.23.1 preprocessing.

The missing edge appears to be between the PJRT execution context and the ROCm host API launch lines. The existing launch-to-device correlation is intact.

**Proposal:**

- Carry the explicit step root through PJRT execution to ROCm API launch XEvents, then let `tsl/profiler/utils/group_events.cc` propagate it through the existing correlation IDs.
- Do **not** infer `one HLO module = one step`: a framework step can execute multiple modules, and a module can execute multiple times per step.

**Desired behaviour:** The capture produces 11 device steps, a populated Steps derived line, and non-zero Overview/Input Pipeline step tables.

#### 4.2 Populate complete-step GPU HLO metrics

**Dependency.** Complete explicit-step propagation first.

**Issue:** `hlo_metrics_db_complete_steps_only` is combined only inside the TPU branch of `xprof/convert/xplane_to_op_stats.cc`. Step propagation alone will not populate it.

**Proposal:** Combine the retained, non-overlapping GPU step HLO databases after incomplete-step filtering. Support the active `FlatOpMetricsDb` migration rather than extending only the legacy proto.

**Desired behaviour:** GPU traces with valid steps contain a non-empty complete-step HLO database and an average-step Roofline record.

#### 4.3 Recognize BF16 in precision statistics

**Dependency.** Complete explicit-step propagation first.

**Issue:** `ClassifyGpuCompute` recognizes only `half`/`fp16`; BF16 kernels or tensor-shape strings are counted as 32-bit once step grouping is available. The current 0%/0% display is primarily caused by the missing device steps.

<img width="864" height="891" alt="Image" src="https://github.com/user-attachments/assets/9f5e4c9e-27cc-41bf-a4ab-51609b75b40c" />

**Proposal:** Recognize `bf16`/`bfloat16` in both event names and tensor-shape metadata. Do not reclassify TF32 without first defining the UI's intended storage-vs-compute semantics.

**Desired behaviour:** Grouped BF16 workloads contribute to the 16-bit precision total.

### 5. Add HIP graph-node attribution

**Issue:** ROCm events have device, queue/stream, scope, and correlation IDs, but no stable graph-exec/node identity. This weakens attribution for XLA command buffers and hides in-graph copies behind blit-kernel names.

**Proposal:** Consume ROCprofiler HIP-graph attribution and emit presence-aware, vendor-neutral graph exec/node stats. Join these IDs to XLA command-buffer annotations and node kinds. CUDA context-ID parity is not a requirement, and this work is unrelated to XProf's HLO Graph Viewer.

**Desired behaviour:** Repeated launches of one HIP graph retain stable node attribution, while unavailable attribution remains distinguishable from node ID zero.

### 6. Emit accurate theoretical occupancy

This is in progress at https://github.com/ROCm/xla/pull/1120. Currently waiting for a rocprof-sdk provided function to avoid maintaining the occupancy calculation within XLA.

**Issue:** ROCm emits `occupancy_pct=0` for every kernel.

<img width="1866" height="779" alt="Image" src="https://github.com/user-attachments/assets/de8feac5-4efe-400a-bb8b-bdb3edf33b23" />

**Proposal:** Prefer an ROCprofiler/HIP-owned calculator, with XLA serializing its result. Any local fallback must model allocation granularity, SGPR/VGPR/AGPR limits, whole-workgroup quantization, LDS granularity, wave slots, and architecture-specific unified register files. A simple minimum of register, LDS, and wave limits is not sufficient.

**Desired behaviour:** Supported CDNA kernels agree with the reference occupancy calculator; unsupported architectures emit no value rather than a false zero.

### 7. Add optional ROCm hardware-counter sampling

**Issue:** ROCm has no equivalent of the opt-in CUPTI PM sampler, so advanced matrix-pipe, bandwidth, and achieved-occupancy counter tracks are unavailable.

**Proposal:**

- Keep collection explicitly opt-in.
- Prefer a low-perturbation device-wide/sampling mode.
- Do not use dispatch-serialized collection as the default; it changes concurrency and can deadlock co-dependent kernels.
- Define separately how sampled counters, if appropriate, aggregate into Overview's matrix/HBM utilization fields.

**Desired behaviour:** Selected counters render as timestamped XPlane tracks with documented overhead and unsupported-counter behavior.
